File: Chapter1/Codes_Figures/spurious_regression_montecarlo.R
# =========================================================
# Spurious Regression: Monte Carlo Simulations
# =========================================================
# Author:   Alessia Paccagnini
# Textbook: Macroeconometrics
#
# Replicates three figures:
#   1. Monte Carlo evidence (rejection rates, mean R2, mean DW)
#   2. Granger-Newbold rule of thumb (R2 vs DW scatter)
#   3. t-statistic distributions (spurious vs valid)
#
# References:
#   Granger & Newbold (1974). Journal of Econometrics, 2(2), 111-120.
#   Phillips (1986). Journal of Econometrics, 33(3), 311-340.
# =========================================================

library(ggplot2)
library(gridExtra)
library(patchwork)   # for combining ggplots side by side

set.seed(42)

# Output directory: same folder as this script (portable)
out_dir <- tryCatch(dirname(normalizePath(sys.frame(1)$ofile)),
                    error = function(e) getwd())

# --------------------------------------------------------
# Helper functions
# --------------------------------------------------------
dw_stat <- function(resid) sum(diff(resid)^2) / sum(resid^2)

run_regression <- function(T, spurious = TRUE) {
  if (spurious) {
    x <- cumsum(rnorm(T))
    y <- cumsum(rnorm(T))
  } else {
    x <- rnorm(T)
    y <- rnorm(T)
  }
  fit    <- lm(y ~ x)
  s      <- summary(fit)
  list(
    r2       = s$r.squared,
    t_stat   = coef(s)["x", "t value"],
    dw       = dw_stat(residuals(fit)),
    beta_hat = coef(fit)["x"]
  )
}

# =========================================================
# Figure 1: Monte Carlo Evidence
# =========================================================
cat(strrep("=", 60), "\n")
cat("Figure 1: Monte Carlo Evidence on Spurious Regression\n")
cat(strrep("=", 60), "\n")

sample_sizes  <- c(50, 100, 200, 500)
n_simulations <- 1000
critical_value <- 1.96

fig1_stats <- data.frame(
  T               = sample_sizes,
  rejection_rate  = NA_real_,
  mean_r2         = NA_real_,
  mean_dw         = NA_real_
)

for (i in seq_along(sample_sizes)) {
  T <- sample_sizes[i]
  cat(sprintf("  Simulating T = %d...\n", T))

  sims <- replicate(n_simulations, run_regression(T, spurious = TRUE),
                    simplify = FALSE)

  t_stats <- sapply(sims, `[[`, "t_stat")
  r2s     <- sapply(sims, `[[`, "r2")
  dws     <- sapply(sims, `[[`, "dw")

  fig1_stats$rejection_rate[i] <- mean(abs(t_stats) > critical_value) * 100
  fig1_stats$mean_r2[i]        <- mean(r2s)
  fig1_stats$mean_dw[i]        <- mean(dws)

  cat(sprintf("  T=%d: Rejection rate=%.1f%%, Mean R2=%.3f, Mean DW=%.3f\n",
              T, fig1_stats$rejection_rate[i],
              fig1_stats$mean_r2[i], fig1_stats$mean_dw[i]))
}

fig1_stats$T_label <- factor(as.character(fig1_stats$T),
                              levels = as.character(sample_sizes))

# Panel (a)
pa <- ggplot(fig1_stats, aes(x = T_label, y = rejection_rate)) +
  geom_col(fill = "#E74C6F", colour = "white", width = 0.6) +
  geom_hline(yintercept = 5, linetype = "dashed", linewidth = 0.8) +
  geom_text(aes(label = sprintf("%.0f%%", rejection_rate)),
            vjust = -0.5, fontface = "bold", size = 3.8) +
  annotate("text", x = 0.7, y = 6.5, label = "Nominal 5%", size = 3) +
  scale_y_continuous(limits = c(0, 100)) +
  labs(title = "(a) False Rejection Rate at 5% Level",
       x = "Sample Size (T)", y = "Rejection Rate (%)") +
  theme_bw(base_size = 11) +
  theme(plot.title = element_text(face = "bold", size = 12))

# Panel (b)
pb <- ggplot(fig1_stats, aes(x = T_label, y = mean_r2)) +
  geom_col(fill = "#6C9BD1", colour = "white", width = 0.6) +
  scale_y_continuous(limits = c(0, 0.5)) +
  labs(title = expression(bold("(b) Mean " * R^2 * " from Spurious Regressions")),
       x = "Sample Size (T)", y = expression("Mean " * R^2)) +
  theme_bw(base_size = 11) +
  theme(plot.title = element_text(face = "bold", size = 12))

# Panel (c)
pc <- ggplot(fig1_stats, aes(x = T_label, y = mean_dw)) +
  geom_col(fill = "#2E7D32", colour = "white", width = 0.6) +
  geom_hline(yintercept = 2, linetype = "dashed", colour = "red", linewidth = 0.8) +
  annotate("text", x = 0.7, y = 2.1, label = "DW = 2", colour = "red", size = 3) +
  scale_y_continuous(limits = c(0, 2.5)) +
  labs(title = "(c) Mean DW Statistic (Lower = More Autocorrelation)",
       x = "Sample Size (T)", y = "Mean Durbin-Watson") +
  theme_bw(base_size = 11) +
  theme(plot.title = element_text(face = "bold", size = 12))

fig1 <- pa + pb + pc
ggsave(file.path(out_dir, "spurious_regression_montecarlo.png"),
       fig1, width = 16, height = 5, dpi = 300)
ggsave(file.path(out_dir, "spurious_regression_montecarlo.pdf"),
       fig1, width = 16, height = 5)
cat("  -> Saved: spurious_regression_montecarlo.png/pdf\n\n")


# =========================================================
# Figure 2: Granger-Newbold Rule of Thumb
# =========================================================
cat(strrep("=", 60), "\n")
cat("Figure 2: Granger-Newbold Rule of Thumb\n")
cat(strrep("=", 60), "\n")

n_sims_scatter <- 1000
T_scatter      <- 200

cat("  Simulating spurious regressions...\n")
spur <- replicate(n_sims_scatter, run_regression(T_scatter, spurious = TRUE),
                  simplify = FALSE)
spurious_df <- data.frame(
  dw   = sapply(spur, `[[`, "dw"),
  r2   = sapply(spur, `[[`, "r2"),
  type = "Spurious (independent I(1))"
)

cat("  Simulating valid regressions...\n")
valid_list <- lapply(seq_len(n_sims_scatter), function(...) {
  x   <- rnorm(T_scatter)
  y   <- 0.5 * x + rnorm(T_scatter)
  fit <- lm(y ~ x)
  list(r2 = summary(fit)$r.squared, dw = dw_stat(residuals(fit)))
})
valid_df <- data.frame(
  dw   = sapply(valid_list, `[[`, "dw"),
  r2   = sapply(valid_list, `[[`, "r2"),
  type = "Valid (stationary with true relationship)"
)

scatter_df <- rbind(spurious_df, valid_df)

fig2 <- ggplot(scatter_df, aes(x = dw, y = r2, colour = type, shape = type)) +
  # Shaded spurious region
  annotate("ribbon", x = c(0, 1), ymin = c(0, 1), ymax = 1,
           fill = "red", alpha = 0.10) +
  # 45-degree line
  geom_abline(intercept = 0, slope = 1, linetype = "dashed",
              linewidth = 1.2, colour = "black") +
  geom_point(alpha = 0.45, size = 1.8) +
  annotate("text", x = 0.18, y = 0.90,
           label = "Spurious\nRegion\n(R² > DW)",
           colour = "red", fontface = "bold.italic", size = 4.5) +
  scale_colour_manual(values = c("Spurious (independent I(1))"              = "#E74C6F",
                                  "Valid (stationary with true relationship)" = "#5DADE2")) +
  scale_shape_manual(values  = c("Spurious (independent I(1))"              = 16,
                                  "Valid (stationary with true relationship)" = 17)) +
  scale_x_continuous(limits = c(0, 2.5)) +
  scale_y_continuous(limits = c(0, 1.0)) +
  labs(
    title  = expression(bold("Granger-Newbold Rule of Thumb: If " * R^2 * " > DW, Suspect Spurious Regression")),
    x      = "Durbin-Watson Statistic",
    y      = expression(R^2),
    colour = NULL, shape = NULL
  ) +
  theme_bw(base_size = 12) +
  theme(legend.position = "bottom",
        plot.title = element_text(face = "bold", size = 13))

ggsave(file.path(out_dir, "spurious_regression_rule_of_thumb.png"),
       fig2, width = 10, height = 7, dpi = 300)
ggsave(file.path(out_dir, "spurious_regression_rule_of_thumb.pdf"),
       fig2, width = 10, height = 7)
cat("  -> Saved: spurious_regression_rule_of_thumb.png/pdf\n\n")


# =========================================================
# Figure 3: t-statistic distributions
# =========================================================
cat(strrep("=", 60), "\n")
cat("Figure 3: Distribution of t-statistics\n")
cat(strrep("=", 60), "\n")

n_sims_tstat <- 5000
T_tstat      <- 200

cat("  Simulating spurious regressions...\n")
spurious_t <- replicate(n_sims_tstat,
                         run_regression(T_tstat, spurious = TRUE)$t_stat)

cat("  Simulating valid regressions...\n")
valid_t <- replicate(n_sims_tstat,
                      run_regression(T_tstat, spurious = FALSE)$t_stat)

# Theoretical t-density
t_grid <- seq(-15, 15, length.out = 500)
theory_df <- data.frame(t = t_grid, density = dt(t_grid, df = T_tstat - 2))

make_tstat_panel <- function(t_vals, fill_col, panel_label, subtitle) {
  df <- data.frame(t = t_vals)
  ggplot(df, aes(x = t)) +
    geom_histogram(aes(y = after_stat(density)), bins = 50,
                   fill = fill_col, alpha = 0.6, colour = "white") +
    geom_line(data = theory_df, aes(x = t, y = density),
              linetype = "dashed", linewidth = 0.9, colour = "black") +
    geom_vline(xintercept = c(-1.96, 1.96),
               linetype = "dotted", linewidth = 0.8, colour = "blue") +
    scale_x_continuous(limits = c(-15, 15)) +
    scale_y_continuous(limits = c(0, 0.40)) +
    labs(title    = panel_label,
         subtitle = subtitle,
         x        = "t-statistic",
         y        = "Density") +
    theme_bw(base_size = 11) +
    theme(plot.title    = element_text(face = "bold", size = 12),
          plot.subtitle = element_text(size = 10))
}

p3a <- make_tstat_panel(spurious_t, "#E74C6F",
                         "(a) t-statistics from Spurious Regressions",
                         "Two Independent Random Walks")
p3b <- make_tstat_panel(valid_t,    "#5DADE2",
                         "(b) t-statistics from Valid Regressions",
                         "Two Independent White Noise Series")

fig3 <- p3a + p3b
ggsave(file.path(out_dir, "spurious_regression_tstat_distribution.png"),
       fig3, width = 14, height = 5.5, dpi = 300)
ggsave(file.path(out_dir, "spurious_regression_tstat_distribution.pdf"),
       fig3, width = 14, height = 5.5)
cat("  -> Saved: spurious_regression_tstat_distribution.png/pdf\n\n")

cat(strrep("=", 60), "\n")
cat("All figures generated successfully!\n")
cat(strrep("=", 60), "\n")

File: Chapter1/Codes_Figures/spurious_regression_montecarlo.py
"""
Spurious Regression: Monte Carlo Simulations
=============================================
Replicates the three figures from the macroeconometrics textbook:
1. Monte Carlo evidence (rejection rates, mean R2, mean DW)
2. Granger-Newbold rule of thumb (R2 vs DW scatter)
3. t-statistic distributions (spurious vs valid)

Author:   Alessia Paccagnini
Textbook: Macroeconometrics

References:
- Granger, C.W.J. and Newbold, P. (1974). "Spurious Regressions in Econometrics."
  Journal of Econometrics, 2(2), 111-120.
- Phillips, P.C.B. (1986). "Understanding Spurious Regressions in Econometrics."
  Journal of Econometrics, 33(3), 311-340.
"""

import numpy as np
import matplotlib.pyplot as plt
import statsmodels.api as sm
from statsmodels.stats.stattools import durbin_watson
from scipy import stats
import os

np.random.seed(42)

# Output directory: same folder as this script (portable across machines)
out_dir = os.path.dirname(os.path.abspath(__file__)) if '__file__' in dir() else '.'

# ============================================================
# Helper function: run one spurious or valid regression
# ============================================================
def run_regression(T, spurious=True):
    """
    Run a single regression of y on x.
    If spurious=True: both x and y are independent random walks (I(1)).
    If spurious=False: both x and y are independent white noise (I(0)).
    
    Returns: dict with r2, t_stat, dw, beta_hat
    """
    if spurious:
        x = np.cumsum(np.random.randn(T))
        y = np.cumsum(np.random.randn(T))
    else:
        x = np.random.randn(T)
        y = np.random.randn(T)
    
    X = sm.add_constant(x)
    model = sm.OLS(y, X).fit()
    
    return {
        'r2': model.rsquared,
        't_stat': model.tvalues[1],
        'dw': durbin_watson(model.resid),
        'beta_hat': model.params[1]
    }


# ============================================================
# Figure 1: Monte Carlo evidence (rejection rates, mean R2, mean DW)
# ============================================================
print("=" * 60)
print("Figure 1: Monte Carlo Evidence on Spurious Regression")
print("=" * 60)

sample_sizes = [50, 100, 200, 500]
n_simulations = 1000
alpha = 0.05
critical_value = 1.96

mc_results = {T: [] for T in sample_sizes}

for T in sample_sizes:
    print(f"  Simulating T = {T}...")
    for _ in range(n_simulations):
        mc_results[T].append(run_regression(T, spurious=True))

# Compute summary statistics
rejection_rates = []
mean_r2 = []
mean_dw = []

for T in sample_sizes:
    t_stats = [r['t_stat'] for r in mc_results[T]]
    r2s = [r['r2'] for r in mc_results[T]]
    dws = [r['dw'] for r in mc_results[T]]
    
    rej_rate = np.mean([abs(t) > critical_value for t in t_stats]) * 100
    rejection_rates.append(rej_rate)
    mean_r2.append(np.mean(r2s))
    mean_dw.append(np.mean(dws))
    
    print(f"  T={T}: Rejection rate={rej_rate:.1f}%, "
          f"Mean R2={np.mean(r2s):.3f}, Mean DW={np.mean(dws):.3f}")

# Plot Figure 1
fig, axes = plt.subplots(1, 3, figsize=(16, 5))

# Panel (a): Rejection rates
bars_a = axes[0].bar([str(T) for T in sample_sizes], rejection_rates, 
                      color='#E74C6F', edgecolor='white', width=0.6)
axes[0].axhline(y=5, color='black', linestyle='--', linewidth=1.5, label='Nominal 5% level')
axes[0].set_xlabel('Sample Size (T)', fontsize=11)
axes[0].set_ylabel('Rejection Rate (%)', fontsize=11)
axes[0].set_title('(a) False Rejection Rate at 5% Level', fontsize=12, fontweight='bold')
axes[0].set_ylim(0, 100)
axes[0].legend(fontsize=10)
for bar, rate in zip(bars_a, rejection_rates):
    axes[0].text(bar.get_x() + bar.get_width()/2, bar.get_height() + 1.5,
                 f'{rate:.0f}%', ha='center', va='bottom', fontweight='bold', fontsize=11)

# Panel (b): Mean R2
bars_b = axes[1].bar([str(T) for T in sample_sizes], mean_r2, 
                      color='#6C9BD1', edgecolor='white', width=0.6)
axes[1].set_xlabel('Sample Size (T)', fontsize=11)
axes[1].set_ylabel('Mean $R^2$', fontsize=11)
axes[1].set_title('(b) Mean $R^2$ from Spurious Regressions', fontsize=12, fontweight='bold')
axes[1].set_ylim(0, 0.5)

# Panel (c): Mean DW
bars_c = axes[2].bar([str(T) for T in sample_sizes], mean_dw, 
                      color='#2E7D32', edgecolor='white', width=0.6)
axes[2].axhline(y=2.0, color='red', linestyle='--', linewidth=1.5, label='DW = 2 (no autocorrelation)')
axes[2].set_xlabel('Sample Size (T)', fontsize=11)
axes[2].set_ylabel('Mean Durbin-Watson', fontsize=11)
axes[2].set_title('(c) Mean DW Statistic (Lower = More Autocorrelation)', fontsize=12, fontweight='bold')
axes[2].set_ylim(0, 2.5)
axes[2].legend(fontsize=9)

plt.tight_layout()
plt.savefig(os.path.join(out_dir, 'spurious_regression_montecarlo.png'), dpi=300, bbox_inches='tight')
plt.savefig(os.path.join(out_dir, 'spurious_regression_montecarlo.pdf'), bbox_inches='tight')
plt.close()
print("  -> Saved: spurious_regression_montecarlo.png/pdf\n")


# ============================================================
# Figure 2: Granger-Newbold Rule of Thumb (R2 vs DW scatter)
# ============================================================
print("=" * 60)
print("Figure 2: Granger-Newbold Rule of Thumb")
print("=" * 60)

n_sims_scatter = 1000
T_scatter = 200

# Spurious regressions (independent random walks)
print("  Simulating spurious regressions...")
spurious_results = [run_regression(T_scatter, spurious=True) for _ in range(n_sims_scatter)]
spurious_r2 = [r['r2'] for r in spurious_results]
spurious_dw = [r['dw'] for r in spurious_results]

# Valid regressions (stationary with true relationship)
print("  Simulating valid regressions...")
valid_r2 = []
valid_dw = []
for _ in range(n_sims_scatter):
    x = np.random.randn(T_scatter)
    y = 0.5 * x + np.random.randn(T_scatter)  # True relationship
    X = sm.add_constant(x)
    model = sm.OLS(y, X).fit()
    valid_r2.append(model.rsquared)
    valid_dw.append(durbin_watson(model.resid))

# Plot Figure 2
fig, ax = plt.subplots(figsize=(10, 7))

# Shaded spurious region (R2 > DW)
dw_fill = np.linspace(0, 2.5, 300)
ax.fill_between(dw_fill, dw_fill, 1.0, alpha=0.15, color='red')
ax.text(0.12, 0.82, 'Spurious\nRegion\n($R^2 > DW$)', fontsize=14, 
        color='red', fontweight='bold', fontstyle='italic', transform=ax.transAxes)

# 45-degree line: R2 = DW
dw_line = np.linspace(0, 1.0, 100)
ax.plot(dw_line, dw_line, 'k--', linewidth=2.5, label='$R^2$ = DW line')

# Scatter: spurious
ax.scatter(spurious_dw, spurious_r2, color='#E74C6F', alpha=0.5, s=30, 
           edgecolors='#C0392B', linewidth=0.3, label='Spurious (independent I(1))')

# Scatter: valid
ax.scatter(valid_dw, valid_r2, color='#5DADE2', alpha=0.4, s=30, marker='^',
           edgecolors='#2874A6', linewidth=0.3, label='Valid (stationary with true relationship)')

ax.set_xlabel('Durbin-Watson Statistic', fontsize=12)
ax.set_ylabel('$R^2$', fontsize=12)
ax.set_title("Granger-Newbold Rule of Thumb: If $R^2 > DW$, Suspect Spurious Regression",
             fontsize=13, fontweight='bold')
ax.set_xlim(0, 2.5)
ax.set_ylim(0, 1.0)
ax.legend(fontsize=11, loc='lower right')

plt.tight_layout()
plt.savefig(os.path.join(out_dir, 'spurious_regression_rule_of_thumb.png'), dpi=300, bbox_inches='tight')
plt.savefig(os.path.join(out_dir, 'spurious_regression_rule_of_thumb.pdf'), bbox_inches='tight')
plt.close()
print("  -> Saved: spurious_regression_rule_of_thumb.png/pdf\n")


# ============================================================
# Figure 3: t-statistic distributions (spurious vs valid)
# ============================================================
print("=" * 60)
print("Figure 3: Distribution of t-statistics")
print("=" * 60)

n_sims_tstat = 5000
T_tstat = 200

# Spurious t-statistics
print("  Simulating spurious regressions...")
spurious_t = [run_regression(T_tstat, spurious=True)['t_stat'] for _ in range(n_sims_tstat)]

# Valid t-statistics
print("  Simulating valid regressions...")
valid_t = [run_regression(T_tstat, spurious=False)['t_stat'] for _ in range(n_sims_tstat)]

# Theoretical t-distribution
t_grid = np.linspace(-15, 15, 500)
t_pdf = stats.t.pdf(t_grid, df=T_tstat - 2)

# Plot Figure 3
fig, axes = plt.subplots(1, 2, figsize=(14, 5.5))

# Panel (a): Spurious
axes[0].hist(spurious_t, bins=50, density=True, color='#E74C6F', alpha=0.6, 
             edgecolor='white', label='Spurious (I(1) variables)')
axes[0].plot(t_grid, t_pdf, 'k--', linewidth=2, label='Standard $t$-distribution')
axes[0].axvline(x=-1.96, color='blue', linestyle=':', linewidth=1.5)
axes[0].axvline(x=1.96, color='blue', linestyle=':', linewidth=1.5, label='$\\pm 1.96$ critical values')
axes[0].set_xlabel('$t$-statistic', fontsize=11)
axes[0].set_ylabel('Density', fontsize=11)
axes[0].set_title('(a) $t$-statistics from Spurious Regressions\n(Two Independent Random Walks)',
                   fontsize=12, fontweight='bold')
axes[0].set_xlim(-15, 15)
axes[0].set_ylim(0, 0.40)
axes[0].legend(fontsize=10)

# Panel (b): Valid
axes[1].hist(valid_t, bins=50, density=True, color='#5DADE2', alpha=0.6,
             edgecolor='white', label='Valid (stationary variables)')
axes[1].plot(t_grid, t_pdf, 'k--', linewidth=2, label='Standard $t$-distribution')
axes[1].axvline(x=-1.96, color='blue', linestyle=':', linewidth=1.5)
axes[1].axvline(x=1.96, color='blue', linestyle=':', linewidth=1.5, label='$\\pm 1.96$ critical values')
axes[1].set_xlabel('$t$-statistic', fontsize=11)
axes[1].set_ylabel('Density', fontsize=11)
axes[1].set_title('(b) $t$-statistics from Valid Regressions\n(Two Independent White Noise Series)',
                   fontsize=12, fontweight='bold')
axes[1].set_xlim(-15, 15)
axes[1].set_ylim(0, 0.40)
axes[1].legend(fontsize=10)

plt.tight_layout()
plt.savefig(os.path.join(out_dir, 'spurious_regression_tstat_distribution.png'), dpi=300, bbox_inches='tight')
plt.savefig(os.path.join(out_dir, 'spurious_regression_tstat_distribution.pdf'), bbox_inches='tight')
plt.close()
print("  -> Saved: spurious_regression_tstat_distribution.png/pdf\n")

print("=" * 60)
print("All figures generated successfully!")
print("=" * 60)
